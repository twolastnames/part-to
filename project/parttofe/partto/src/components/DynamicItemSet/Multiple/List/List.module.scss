@import "../../../../App.scss";

.list {
  overflow-y: scroll;
}

.item {
  width: 100%;
  cursor: pointer;
  overflow-y: scroll;
  padding: var(--gap-medium) 0;
  font: var(--font-regular-large);
}

.list > div:first-of-type {
  border-top: var(--detail-color) 1px solid;
}

.row {
  display: flex;
  flex-direction: row;
  height: fit-content;
  border-bottom: var(--detail-color) 1px solid;
  margin-right: var(--gap-xlarge);
}
