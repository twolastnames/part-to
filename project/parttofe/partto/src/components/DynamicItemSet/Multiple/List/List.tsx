import React, {
  PropsWithChildren,
  useLayoutEffect,
  useRef,
  useState,
} from "react";

import { Carousel } from "../Carousel/Carousel";
import { ButtonProps } from "../../../Button/ButtonTypes";
import { MultipleProps } from "../MultipleTypes";

interface ShellProps extends PropsWithChildren {
  height: number;
  operations: Array<ButtonProps>;
}

function Shell({ children, height, operations }: ShellProps) {
  return <div style={{ height }}>{children}</div>;
}

export function List({ items }: MultipleProps) {
  const minimumHeight = 120;
  const domRef = useRef<HTMLDivElement>(null);
  const [{ height, perPage }, setCalculated] = useState<{
    height: number;
    perPage: number;
  }>({ height: 0, perPage: 0 });

  const calculateHeight = () => {
    if (!domRef.current) {
      return;
    }
    const { height: rectangleHeight } =
      domRef?.current?.getBoundingClientRect();
    const screenHeight = rectangleHeight;
    const count = Math.floor(screenHeight / minimumHeight);
    const height = Math.floor(screenHeight / count);

    setCalculated({
      height,
      perPage: count === 0 ? 1 : count,
    });
  };

  useLayoutEffect(() => {
    calculateHeight();
  }, [domRef]);

  const pages = [];
  if (height > 0) {
    for (let index = 0; index < items.length; index += perPage) {
      pages.push({
        view: (
          <div key={index}>
            {items.slice(index, index + perPage).map((item, key) => (
              <Shell
                key={`${index}-${key}`}
                height={height}
                operations={item.itemOperations}
              >
                {item.listView}
              </Shell>
            ))}
          </div>
        ),
      });
    }
  }

  return (
    <div data-testid="List" style={{ height: "100%" }} ref={domRef}>
      {items.length > 0 ? <Carousel items={items} /> : null}
    </div>
  );
}
