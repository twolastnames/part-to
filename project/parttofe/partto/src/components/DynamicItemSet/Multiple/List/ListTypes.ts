import { MultipleProps } from "../MultipleTypes";

export interface ListProps extends MultipleProps {
  onSelectionChanged: () => void;
}
