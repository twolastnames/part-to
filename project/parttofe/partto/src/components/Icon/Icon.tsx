import React, { ForwardRefExoticComponent, RefAttributes } from "react";
import { rem } from "@mantine/core";
import {
  Icon as TablerIcon,
  IconProps as TablerIconProps,
  IconPlayerTrackNext,
  IconPlayerPlay,
  IconList,
  IconFile,
  IconCaretDown,
  IconCaretUp,
  IconChefHat,
  IconCooker,
  IconToolsKitchen,
} from "@tabler/icons-react";

import classes from "./Icon.module.scss";

export type IconType = ForwardRefExoticComponent<
  TablerIconProps & RefAttributes<TablerIcon>
>;

export enum Size {
  Small = 20,
  Medium = 30,
  Large = 50,
  ExtraLarge = 80,
}

export interface IconProps {
  definition: IconType;
  size?: Size;
  onClick?: () => void;
}

export function Icon({ onClick, definition, size }: IconProps) {
  return React.createElement(definition, {
    ...(onClick
      ? {
          onClick,
          className: classes.clickable,
        }
      : {}),
    style: {
      width: rem((size || Size.Medium) as number),
      height: rem((size || Size.Medium) as number),
    },
  });
}

export const Next = IconPlayerTrackNext;
export const Start = IconPlayerPlay;
export const List = IconList;
export const File = IconFile;
export const Up = IconCaretUp;
export const Down = IconCaretDown;
export const ChefHat = IconChefHat;
export const Oven = IconCooker;
export const KitchenTools = IconToolsKitchen;
