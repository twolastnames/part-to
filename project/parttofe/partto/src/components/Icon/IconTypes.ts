import {
  Icon as TablerIcon,
  IconProps as TablerIconProps,
} from "@tabler/icons-react";
import { ForwardRefExoticComponent, RefAttributes } from "react";

export type IconType = ForwardRefExoticComponent<
  TablerIconProps & RefAttributes<TablerIcon>
>;

export enum Size {
  Small = 20,
  Medium = 30,
  Large = 50,
  ExtraLarge = 80,
}

export interface IconProps {
  definition: IconType;
  size?: Size;
  onClick?: () => void;
}
