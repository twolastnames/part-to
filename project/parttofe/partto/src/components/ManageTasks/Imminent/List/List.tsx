import React from "react";

import classes from "./List.module.scss";
import { ImminentProps } from "../ImminentTypes";
import { useTaskGet } from "../../../../api/taskget";
import { Spinner } from "../../../Spinner/Spinner";
import { TimerString } from "../TimerString/TimerString";

export function List({ timestamp, till, duty }: ImminentProps) {
  const taskResponse = useTaskGet({ task: duty });
  return (
    <Spinner responses={[taskResponse]}>
      <div className={classes.list} data-testid="List">
        "Duty Start Time:"
        <TimerString started={timestamp} duration={till} />
        {`to "${taskResponse.data?.description}"`}
      </div>
    </Spinner>
  );
}
