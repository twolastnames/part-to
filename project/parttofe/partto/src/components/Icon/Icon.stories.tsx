import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Icon, Next } from "./Icon";
import { ShellProvider } from "../../providers/ShellProvider";
import { Size } from "./IconTypes";

const meta: Meta<typeof Icon> = {
  component: Icon,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Icon>;

export const Medium: Story = {
  args: { definition: Next },
};

export const Small: Story = {
  args: { definition: Next, size: Size.Small },
};

export const Large: Story = {
  args: { definition: Next, size: Size.Large },
};

export const ExtraLarge: Story = {
  args: { definition: Next, size: Size.ExtraLarge },
};
