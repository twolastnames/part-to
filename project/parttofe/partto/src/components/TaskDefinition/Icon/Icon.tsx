import React from "react";

import classes from "./Icon.module.scss";
import { ClassNames, IconProps } from "./IconTypes";
import {
  Duty as DutyIcon,
  Imminent as ImminentIcon,
  Icon as PhysicalIcon,
  Task as TaskIcon,
} from "../../Icon/Icon";

export const Duty: ClassNames = {
  imminent: classes.hidden,
  duty: classes.icon,
  task: classes.hidden,
};

export const Task: ClassNames = {
  imminent: classes.hidden,
  duty: classes.hidden,
  task: classes.icon,
};

export const Imminent: ClassNames = {
  imminent: classes.icon,
  duty: classes.hidden,
  task: classes.hidden,
};

export function Icon({ classNames: { task, duty, imminent } }: IconProps) {
  return (
    <>
      <span className={imminent}>
        <PhysicalIcon definition={ImminentIcon} />
      </span>
      <span className={duty}>
        <PhysicalIcon definition={DutyIcon} />
      </span>
      <span className={task}>
        <PhysicalIcon definition={TaskIcon} />
      </span>
    </>
  );
}
