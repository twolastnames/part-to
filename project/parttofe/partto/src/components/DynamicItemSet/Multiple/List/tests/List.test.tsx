import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../../../providers/ShellProvider";
import { List } from "../List";
import { Next } from "../../../../Icon/Icon";
import { LeftContext } from "../../../../../providers/DynamicItemSetPair";

test("snapshot", async () => {
  render(
    <ShellProvider>
      <List
        context={LeftContext}
        items={[
          {
            key: "9",
            listView: <div>hello</div>,
            detailView: <div>goodbye</div>,
            itemOperations: [
              { text: "hello", icon: Next, onClick: () => undefined },
            ],
          },
          {
            key: "7",
            listView: <div>hello</div>,
            detailView: <div>goodbye</div>,
            itemOperations: [
              { text: "goodbye", icon: Next, onClick: () => undefined },
            ],
          },
        ]}
      />
    </ShellProvider>,
  );
  const component = await screen.findByTestId("List");
  expect(component).toMatchSnapshot();
});
