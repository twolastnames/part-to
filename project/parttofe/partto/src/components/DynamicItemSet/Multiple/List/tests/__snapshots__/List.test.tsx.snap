// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="list"
  data-testid="List"
  style="height: 100%;"
>
  <div
    class="row"
  >
    <div
      class="item"
    >
      <div>
        hello
      </div>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="item"
    >
      <div>
        hello
      </div>
    </div>
  </div>
</div>
`;
